File: be/src/main/java/sg/flow/auth/RefreshTokenAuthentcationFilter.java
package sg.flow.auth;

import java.io.IOException;
import java.util.Optional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter;
import org.springframework.security.web.util.matcher.AntPathRequestMatcher;

import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import sg.flow.models.auth.TokenSet;
import sg.flow.services.AuthServices.FlowTokenService;

public class RefreshTokenAuthentcationFilter extends AbstractAuthenticationProcessingFilter {

    @Autowired
    private final FlowTokenService tokenService;

    public RefreshTokenAuthentcationFilter(FlowTokenService tokenService, AuthenticationManager authManager) {
        super(new AntPathRequestMatcher("/auth/getAccessToken", "POST"));
        setAuthenticationManager(authManager);
        this.tokenService = tokenService;
    }

    @Override
    public Authentication attemptAuthentication(HttpServletRequest request, HttpServletResponse response)
            throws AuthenticationException, IOException, ServletException {
        String header = request.getHeader("Authorization");
        if (header == null || !header.startsWith("Bearer ")) {
            throw new BadCredentialsException("No Bearer token found in request headers");
        }
        String refreshToken = header.substring(7);
        Authentication auth = handleRefreshToken(refreshToken);

        return auth;
    }

    private Authentication handleRefreshToken(String refreshToken) {
        Optional<TokenSet> maybeNewTokenSet = tokenService.getAccessTokenByRefreshToken(refreshToken);
        if (!maybeNewTokenSet.isPresent()) {
            throw new BadCredentialsException("Refresh token invalid");
        }
        TokenSet newtokenSet = maybeNewTokenSet.get();
        String newAccessToken = newtokenSet.getAccessToken();
        String newRefreshToken = newtokenSet.getRefreshToken();

        UsernamePasswordAuthenticationToken authentication = new UsernamePasswordAuthenticationToken(null, null, null);

        authentication.setDetails(new TokenSet(newAccessToken, newRefreshToken));

        return authentication;
    }
}

File: be/src/main/kotlin/sg/flow/auth/RefreshTokenAuthentcationFilter.kt
package sg.flow.auth

import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import kotlinx.coroutines.runBlocking
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.BadCredentialsException
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter
import org.springframework.security.web.util.matcher.AntPathRequestMatcher
import sg.flow.models.auth.TokenSet
import sg.flow.services.AuthServices.FlowTokenService

class RefreshTokenAuthentcationFilter(
        private val tokenService: FlowTokenService,
        authManager: AuthenticationManager,
) : AbstractAuthenticationProcessingFilter(AntPathRequestMatcher("/auth/getAccessToken", "POST")) {

    init {
        setAuthenticationManager(authManager)
    }

    override fun attemptAuthentication(
            request: HttpServletRequest,
            response: HttpServletResponse
    ): Authentication? {
        val header =
                request.getHeader("Authorization")
                        ?: throw BadCredentialsException("No Authorization header")
        if (!header.startsWith("Bearer ")) {
            throw BadCredentialsException("No Bearer token")
        }
        val refreshToken = header.substring(7)
        val auth =
                handleRefreshToken(refreshToken)
                        ?: throw BadCredentialsException("Invalid refresh token")
        return auth
    }

    private fun handleRefreshToken(refreshToken: String): Authentication? = runBlocking {
        val tokenSet =
                tokenService.getAccessTokenByRefreshToken(refreshToken) ?: return@runBlocking null
        val auth = UsernamePasswordAuthenticationToken(null, null, emptyList())
        auth.details = TokenSet(tokenSet.accessToken, tokenSet.refreshToken)
        auth
    }
}
